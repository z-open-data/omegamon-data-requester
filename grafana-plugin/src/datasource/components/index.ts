export { tid } from './tid';
export { CollapseWithInfoIcon } from './CollapseWithInfoIcon';
export { GrafanaThemeProvider } from './GrafanaThemeProvider';
