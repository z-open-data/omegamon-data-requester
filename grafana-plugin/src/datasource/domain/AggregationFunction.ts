import { includes } from 'public-common';

export const ALLOWED_AGGREGATION_FUNCS = ['AVG', 'COUNT', 'MAX', 'MIN', 'SUM'] as const;

export type AggregationFuncName = (typeof ALLOWED_AGGREGATION_FUNCS)[number];

const columnWithAggregationFunctionRegex = /^[^()]+\([^()]+\)$/;

export function fieldNameContainsAggFunc(fieldName: string): boolean {
  return columnWithAggregationFunctionRegex.test(fieldName);
}

export function extractAggFuncFromFieldName(fieldName: string): AggregationFuncName | null {
  if (!fieldNameContainsAggFunc(fieldName)) {
    return null;
  }
  const possibleAggFunc = fieldName.slice(0, fieldName.indexOf('('));
  if (includes(ALLOWED_AGGREGATION_FUNCS, possibleAggFunc)) {
    return possibleAggFunc;
  }
  throw new Error(`Unknown aggregation function ${possibleAggFunc}`);
}

export function extractColumnIdFromFieldName(fieldName: string): string {
  if (!fieldNameContainsAggFunc(fieldName)) {
    return fieldName;
  }
  return fieldName.slice(fieldName.indexOf('(') + 1, fieldName.indexOf(')'));
}
