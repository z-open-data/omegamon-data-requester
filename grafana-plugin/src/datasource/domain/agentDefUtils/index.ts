export * from './parseAgentDef';
export * from './stringifyAgentDef';
