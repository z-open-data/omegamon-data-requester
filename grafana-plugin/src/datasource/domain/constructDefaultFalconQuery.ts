import { DataQuery } from '@grafana/schema';
import { DiscriminatedUnionMap } from 'public-common';
import { AffinityId } from 'public-domain';

import {
  FalconQuery,
  FALCON_QUERY_VERSION,
  MetricsQueryParams,
  FalconMetricsQuery,
  FalconSituationsQuery,
  FalconQueryType,
} from './FalconQuery';

const defaultCommonQueryParams: Pick<FalconQuery, 'falconVersion' | 'hide' | 'refId'> = {
  falconVersion: FALCON_QUERY_VERSION,
  refId: 'A',
  hide: false,
};

export const defaultMetricsQueryParams: MetricsQueryParams = {
  affinityId: '' as AffinityId,
  tableId: '',
  columns: [],
  agentsAndGroups: [],
  history: false,
  filter: {},
  orderBy: [],
  groupBy: [],
  parmas: [],
};

const defaultMetricsQuery: FalconMetricsQuery = {
  queryType: 'metrics',
  ...defaultCommonQueryParams,
  falconParams: defaultMetricsQueryParams,
};

const defaultSituationsQuery: FalconSituationsQuery = { queryType: 'situations', ...defaultCommonQueryParams };

const defaultQueryMap: DiscriminatedUnionMap<FalconQuery, 'queryType'> = {
  metrics: defaultMetricsQuery,
  situations: defaultSituationsQuery,
};

export function constructDefaultFalconQuery(queryType: FalconQueryType, previousQuery?: DataQuery): FalconQuery {
  return {
    ...defaultQueryMap[queryType],
    refId: previousQuery?.refId ?? defaultQueryMap[queryType].refId,
    hide: previousQuery?.hide ?? defaultQueryMap[queryType].hide,
    key: previousQuery?.key,
    datasource: previousQuery?.datasource,
  };
}
