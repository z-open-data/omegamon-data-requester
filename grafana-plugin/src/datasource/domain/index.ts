export type { FalconDatasourceJsonData, FalconDatasourceConfig, UnknownConfig } from './FalconDatasourceConfig';
export { FALCON_DATASOURCE_OPTIONS_VERSION, constructDefaultConfig } from './FalconDatasourceConfig';
export { ALLOWED_COMPARISON_OPERATORS, EMPTY_PARMA, FALCON_QUERY_VERSION } from './FalconQuery';
export type {
  FalconQuery,
  FalconQueryType,
  FalconSituationsQuery,
  FalconMetricsQuery,
  ComparisonOperator,
  MetricsQueryFilters,
  MetricsQueryColumn,
  MetricsQueryFilter,
  MetricsQueryFilterOf,
  MetricsQueryFilterClause,
  MetricsQueryOrderByItem,
  MetricsQueryParams,
  MetricsQueryParma,
  SourceDef,
  AgentOrGroupName,
  AtLparDef,
  BaseQueryFilters,
  BaseQueryFilter,
} from './FalconQuery';

export { constructDefaultFalconQuery, defaultMetricsQueryParams } from './constructDefaultFalconQuery';

export type { AggregationFuncName } from './AggregationFunction';
export {
  ALLOWED_AGGREGATION_FUNCS,
  extractAggFuncFromFieldName,
  fieldNameContainsAggFunc,
  extractColumnIdFromFieldName,
} from './AggregationFunction';
