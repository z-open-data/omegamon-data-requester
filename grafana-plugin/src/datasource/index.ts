// eslint-disable-next-line import/no-cycle
export { FalconDatasource } from './datasource';
