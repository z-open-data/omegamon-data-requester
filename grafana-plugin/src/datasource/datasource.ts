import {
  DataQueryResponse,
  DataQueryErrorType,
  ScopedVars,
  DataQueryRequest,
  DataSourceInstanceSettings,
  MetricFindValue,
} from '@grafana/data';
import { DataSourceWithBackend } from '@grafana/runtime';
import { isEqual, uniq } from 'lodash';
import { combineWithAnd, errorToString, throwIfNullish, unwrapObservable } from 'public-common';
import { TableMetadata } from 'public-domain';
import { from, Observable, switchMap, tap } from 'rxjs';

import {
  FalconQuery,
  FalconDatasourceJsonData,
  EMPTY_PARMA,
  MetricsQueryParma,
  FalconMetricsQuery,
  MetricsQueryFilter,
  FalconSituationsQuery,
  MetricsQueryFilterClause,
} from 'datasource/domain';
import { formatMetricsInResponse } from 'datasource/features/formatting/format';
import { ItmQueryParams, convertQueryToItmFormat } from 'datasource/features/formatting/reverseFormat';
import { MetadataLoader } from 'datasource/features/metadata';
import { addTakeActionToResponse } from 'datasource/features/takeAction';
import { validateMetricsQuery, ValidationProblem } from 'datasource/features/validation/validateMetricsQuery';
// eslint-disable-next-line import/no-cycle
import {
  applyVariablesInFilters,
  generateFilterFromAdHocs,
  VariableSupport,
  getDatasourceAdHocsFilters,
  getTableIdsFromAdHocs,
  applyVariablesInAgentsAndGroups,
  applyVariablesInParmas,
} from 'datasource/features/variables';
import {
  getRequestWithLatestVersionFalconQueries,
  isFalconQueryOfAnyVersion,
} from 'datasource/features/versioning/falconQuery';
import { AdHocVariableFilter } from 'datasource/grafana';

type PreparedFalconMetricsQuery = Omit<FalconMetricsQuery, 'falconParams'> & { falconParams: ItmQueryParams };
type PreparedDataQuery = PreparedFalconMetricsQuery | FalconSituationsQuery;

export class FalconDatasource extends DataSourceWithBackend<FalconQuery, FalconDatasourceJsonData> {
  private static getErrorResponseObservable(error: unknown): Observable<DataQueryResponse> {
    return new Observable<DataQueryResponse>((observer) => {
      observer.next({
        error: {
          type: DataQueryErrorType.Cancelled,
          message: errorToString(error),
        },
        data: [],
      });
      observer.complete();
    });
  }

  private static applyVariablesInAgentsAndGroups(
    query: FalconMetricsQuery,
    scopedVars: ScopedVars
  ): FalconMetricsQuery {
    const {
      falconParams: { affinityId, agentsAndGroups },
    } = query;

    const processedAgentsAndGroups = applyVariablesInAgentsAndGroups(agentsAndGroups, scopedVars, affinityId);

    const processedQuery: FalconMetricsQuery = {
      ...query,
      falconParams: {
        ...query.falconParams,
        agentsAndGroups: processedAgentsAndGroups,
      },
    };

    return processedQuery;
  }

  private static prepareParmas(query: FalconMetricsQuery, scopedVars: ScopedVars): FalconMetricsQuery {
    const nonEmptyParmas = query.falconParams.parmas.filter((parma) => !isEqual(parma, EMPTY_PARMA));

    const processedParmas: MetricsQueryParma[] = applyVariablesInParmas(nonEmptyParmas, scopedVars);

    const processedQuery: FalconMetricsQuery = {
      ...query,
      falconParams: {
        ...query.falconParams,
        parmas: processedParmas,
      },
    };

    return processedQuery;
  }

  private static applyVariablesInFilters(
    query: FalconMetricsQuery,
    tableMd: TableMetadata,
    scopedVars: ScopedVars
  ): FalconMetricsQuery {
    const filter = applyVariablesInFilters(query.falconParams.filter, tableMd, scopedVars);

    const processedQuery: FalconMetricsQuery = {
      ...query,
      falconParams: {
        ...query.falconParams,
        filter,
      },
    };

    return processedQuery;
  }

  private static applyFiltersFromAdHocs(
    query: FalconMetricsQuery,
    tableMd: TableMetadata,
    adHocVariableFilters: AdHocVariableFilter[]
  ): FalconMetricsQuery {
    const filterFromAdHocs = generateFilterFromAdHocs(adHocVariableFilters, tableMd);

    const nonEmptyFilters = [query.falconParams.filter.nonAggregated, filterFromAdHocs].filter(
      (f): f is MetricsQueryFilter => !!f
    );

    const combined = combineWithAnd<MetricsQueryFilterClause>(...nonEmptyFilters);

    const nonAggregated = combined === 'no_filter' ? undefined : combined;

    const processedQuery: FalconMetricsQuery = {
      ...query,
      falconParams: {
        ...query.falconParams,
        filter: {
          ...query.falconParams.filter,
          nonAggregated,
        },
      },
    };

    return processedQuery;
  }

  private static validateIfQueriesInitialized(request: DataQueryRequest<FalconQuery>): void {
    request.targets.forEach((query) => {
      if (!isFalconQueryOfAnyVersion(query)) {
        throw new Error('Query not initialized');
      }
    });
  }

  readonly metadataLoader;
  override readonly variables = new VariableSupport(this);

  constructor(public instanceSettings: DataSourceInstanceSettings<FalconDatasourceJsonData>) {
    super(instanceSettings);
    this.metadataLoader = new MetadataLoader(this);
  }

  private static prepareMetricsQuery(
    query: FalconMetricsQuery,
    tableMd: TableMetadata,
    adHocVariableFilters: AdHocVariableFilter[],
    scopedVars: ScopedVars,
    isVariableRequest: boolean
  ): PreparedFalconMetricsQuery {
    const applyedVariablesQuery = FalconDatasource.applyVariablesInAgentsAndGroups(query, scopedVars);

    const queryWithPreparedParmas = FalconDatasource.prepareParmas(applyedVariablesQuery, scopedVars);

    const queryWithPreparedFilters = FalconDatasource.applyVariablesInFilters(
      queryWithPreparedParmas,
      tableMd,
      scopedVars
    );

    if (isVariableRequest) {
      return {
        ...queryWithPreparedFilters,
        falconParams: convertQueryToItmFormat(queryWithPreparedFilters.falconParams, tableMd),
      };
    }

    const queryWithApplyedAdhocs = FalconDatasource.applyFiltersFromAdHocs(
      queryWithPreparedFilters,
      tableMd,
      adHocVariableFilters
    );

    return {
      ...queryWithApplyedAdhocs,
      falconParams: convertQueryToItmFormat(queryWithApplyedAdhocs.falconParams, tableMd),
    };
  }

  /**
   * Removes empty parmas, applies ad hoc filters, applies variables to filters, reverse formats filter values
   */
  private prepareMetricsQueries(
    request: DataQueryRequest<FalconQuery>,
    tableMetadatas: TableMetadata[],
    isVariableRequest: boolean
  ): DataQueryRequest<PreparedDataQuery> {
    const adHocs = getDatasourceAdHocsFilters(this.uid);

    const preparedQueries = request.targets.map((query) => {
      if (query.queryType === 'situations') {
        return query;
      }

      const tableMd = tableMetadatas.find((tMd) => tMd.id === query.falconParams.tableId);

      throwIfNullish(tableMd, `Provided tableMetadatas must contain table metadata for every metrics query in request`);

      const preparedMetricsQuery = FalconDatasource.prepareMetricsQuery(
        query,
        tableMd,
        adHocs,
        request.scopedVars,
        isVariableRequest
      );

      return preparedMetricsQuery;
    });

    const preparedRequest: DataQueryRequest<PreparedDataQuery> = { ...request, targets: preparedQueries };

    return preparedRequest;
  }

  private validateMetricsQueries = async (request: DataQueryRequest<FalconQuery>): Promise<void> => {
    const validationProblems = await request.targets.reduce(
      async (prev, curr) => {
        if (curr.queryType === 'metrics') {
          const queryValidationProblems = await validateMetricsQuery(curr.falconParams, this.metadataLoader);
          if (queryValidationProblems.length) {
            const allProblems = await prev;
            allProblems.push(...queryValidationProblems);
            return allProblems;
          }
        }
        return prev;
      },
      Promise.resolve([] as ValidationProblem[])
    );

    const validationErrors = validationProblems.filter((problem) => problem.severity === 'error');

    if (validationErrors.length) {
      const errorsString = validationErrors.map(({ message }) => `"${message}"`).join(', ');
      throw new Error(`Found problems in query: ${errorsString}`);
    }
  };

  private async getMetricsQueriesTableMetadatas(request: DataQueryRequest<FalconQuery>): Promise<TableMetadata[]> {
    const queriesTableIds = uniq(
      request.targets
        .filter((query): query is FalconMetricsQuery => query.queryType === 'metrics')
        .map((query) => query.falconParams.tableId)
    );

    const tableMetadatas = await this.metadataLoader.getTableMetadatas(queriesTableIds);

    return tableMetadatas;
  }

  private async queryAndGetFormattedResponseObservable(
    request: DataQueryRequest<FalconQuery>,
    isVariableRequest: boolean,
    panelType?: string
  ): Promise<Observable<DataQueryResponse>> {
    try {
      FalconDatasource.validateIfQueriesInitialized(request);
      const latestVersionRequest = await getRequestWithLatestVersionFalconQueries(request, this.metadataLoader);

      await this.validateMetricsQueries(latestVersionRequest);

      const metricsQueriesTableMetadatas = await this.getMetricsQueriesTableMetadatas(latestVersionRequest);

      const metricsQueryRequest = this.prepareMetricsQueries(
        latestVersionRequest,
        metricsQueriesTableMetadatas,
        isVariableRequest
      );

      /**
       * FYI: typecasting
       * We modify query object before we send it to datasource backend (to resolve formatting, enums, etc).
       * Grafana doesn't allow us to define those 2 different types so 'super.query' takes FalconQuery type.
       * Because of that, we just going to cast it
       */
      const queryRequest = metricsQueryRequest as DataQueryRequest<FalconMetricsQuery>;

      /**
       * FYI: `super.query`:
       * On internet connection failure or server's 400/500 HTTP status response, etc
       * does not throw error, but rather still returns observable,
       * that emits a value (it emits value, not error!)
       */
      const responseObservable = super.query(queryRequest);

      const formattedResponseObservable = await responseObservable.pipe(
        tap((response) => {
          // FYI: mutates response
          formatMetricsInResponse(response, metricsQueriesTableMetadatas);
        }),
        switchMap((response) =>
          from(
            addTakeActionToResponse(
              response,
              request,
              this.metadataLoader,
              metricsQueriesTableMetadatas,
              this.instanceSettings.jsonData.adminUrl ?? this.instanceSettings.url ?? '',
              panelType
            )
          )
        )
      );

      return formattedResponseObservable;
    } catch (err) {
      return FalconDatasource.getErrorResponseObservable(err);
    }
  }

  override query(request: DataQueryRequest<FalconQuery>, isVariableRequest = false): Observable<DataQueryResponse> {
    /**
     * FYI: Why is the code below so confusing (observable inside promise inside observable)?
     * 1. Grafana's `super.query` returns Observable.
     * 2. Observables are awkward to work with.
     * 3. `metrics` queries each provide single response result.
     * 4. `situations` queries each provide multiple response results throughout time
     *    (it uses streaming through websockets).
     * 5. Need to handle request that contains both `metrics` and `situations` queries at same time.
     */
    return unwrapObservable(
      this.queryAndGetFormattedResponseObservable(request, isVariableRequest, request.panelPluginId)
    );
  }

  override async getTagKeys(): Promise<MetricFindValue[]> {
    const tableIds = getTableIdsFromAdHocs(this.uid);
    const columnNames = await this.metadataLoader.getColumnNamesFromTableIds(tableIds);
    // getTagKeys is called once per variable/datasource so I'm OK with sort function here
    return columnNames.sort().map((columnName) => ({ text: columnName }));
  }

  // Grafana's function is overriden and any comes from it
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  override async getTagValues(options: any): Promise<MetricFindValue[]> {
    const tableIds = getTableIdsFromAdHocs(this.uid);
    const columns = await this.metadataLoader.getColumnMetadataForAdHocFilters(tableIds);
    const columnName = options.key as string;
    const { enums = {} } = Object.values(columns).find(({ name }) => columnName === name) ?? {};
    return Object.entries(enums).map(([value, text]) => ({
      value, // This field is ignored by Grafana >_<
      text: `${text}`,
    }));
  }
}
