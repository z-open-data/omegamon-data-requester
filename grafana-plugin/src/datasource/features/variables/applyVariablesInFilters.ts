import { ScopedVars } from '@grafana/data';
import { FilterCombinationResult, throwIfNullish } from 'public-common';
import { TableMetadata, ColumnMetadata } from 'public-domain';

import { MetricsQueryFilters, MetricsQueryFilterClause, MetricsQueryFilter } from 'datasource/domain';
import { transformMetricsQueryFilters } from 'datasource/domain/filters';

import { TemplateVariablesDetails } from './TemplateVariablesDetails';

/**
 * Throws if any template in filters contains variables that are not found in dashboard.
 *
 * Applies variables,
 * potentially changes filters structure (if e.g. some clause uses variable that has multiple values selected or ALL value selected),
 * reverse-formats values of those clauses that had variables applied.
 */
export function applyVariablesInFilters(
  metricsQueryFilters: MetricsQueryFilters,
  tableMd: TableMetadata,
  scopedVars: ScopedVars
): MetricsQueryFilters {
  const transformer = (clause: MetricsQueryFilterClause) => {
    const columnMd = tableMd.columns[clause.columnId];
    throwIfNullish(columnMd, `Table metadata should contain at least column metadata for the column in the clause`);
    return handleClause(clause, columnMd, scopedVars);
  };
  return transformMetricsQueryFilters(metricsQueryFilters, transformer);
}

function handleClause(
  clause: MetricsQueryFilterClause,
  columnMd: ColumnMetadata,
  scopedVars: ScopedVars
): FilterCombinationResult<MetricsQueryFilterClause> {
  const template = clause.userDefinedValue.toString();
  const templateVariablesDetails = new TemplateVariablesDetails(template, scopedVars);

  if (!templateVariablesDetails.containsVariables()) {
    return { clause };
  }

  if (templateVariablesDetails.hasUnspecifiedAllSelected()) {
    if (!templateVariablesDetails.isSingleVariableExpression()) {
      throw new Error(
        `The provided template "${template}" does not match the expected format for variables with value "All"`
      );
    }
    return 'always_true';
  }

  const filledTemplateVariations = templateVariablesDetails.getVariations();

  if (filledTemplateVariations.length > 1) {
    return handleClauseWithMultipleValues(clause, filledTemplateVariations);
  }

  const processedClause: MetricsQueryFilterClause = {
    ...clause,
    userDefinedValue: filledTemplateVariations[0],
  };
  return { clause: processedClause };
}

function handleClauseWithMultipleValues(clause: MetricsQueryFilterClause, varValues: string[]): MetricsQueryFilter {
  const multivalueClauses = varValues.map((value) => {
    const processedClause = { ...clause, userDefinedValue: value };
    const processedFilter: MetricsQueryFilter = { clause: processedClause };
    return processedFilter;
  });
  return { or: multivalueClauses };
}

export const testExports = {
  handleClause,
  handleClauseWithMultipleValues,
};
