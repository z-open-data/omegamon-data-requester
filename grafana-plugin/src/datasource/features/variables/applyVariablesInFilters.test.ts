import { ScopedVars } from '@grafana/data';
import { VariableInterpolation } from '@grafana/runtime';
import { ColumnMetadata } from 'public-domain';

import { MetricsQueryFilters } from 'datasource/domain';

import { testExports, applyVariablesInFilters } from './applyVariablesInFilters';
import {
  compoundFilter,
  noVarInTemplateFilter,
  multiVariablesTemplateFilter,
  allValFilter,
  allInAndFilter,
  allInMultipleClauseAndFilter,
  allInOrFilter,
  complexAllFilter,
  customAllFilter,
  invalidTemplateAllFilter,
  varNotFoundInDashboardFilter,
  mgtstatFilter,
  swpstatFilter,
  frmsusepctFilter,
  singleVarFilter,
  mgtstatClause,
  swpstatClause,
  frmsusepctClause,
  multiValueClause,
  mockedTableMetadata,
  mockedReplaceFnInputOutputMap,
  scopedVars,
  mockedTypedVariableModels,
} from './mocks';

const { handleClause, handleClauseWithMultipleValues } = testExports;

jest.mock('@grafana/runtime', () => ({
  getTemplateSrv: () => ({
    replace: (template: string, scopedVars: ScopedVars, format: string, interpolations: VariableInterpolation[]) => {
      const foundInterpolations = mockedReplaceFnInputOutputMap[template];
      interpolations.push(...(foundInterpolations || []));
      return template;
    },
    getVariables: () => mockedTypedVariableModels,
  }),
}));

describe('applyVariablesInFilters', () => {
  it('should return processed filter with nonAggregated field', () => {
    expect(applyVariablesInFilters(compoundFilter, mockedTableMetadata, scopedVars)).toMatchSnapshot();
  });
  it('should return empty filter when empty filter passed', () => {
    const emptyFilter: MetricsQueryFilters = {};
    expect(applyVariablesInFilters(emptyFilter, mockedTableMetadata, scopedVars)).toEqual(emptyFilter);
  });
  it('should return unchanged filter when there is no variable defined in "userDefinedValue"', () => {
    expect(applyVariablesInFilters(noVarInTemplateFilter, mockedTableMetadata, scopedVars)).toMatchSnapshot();
  });
  it('should return filter with correctly applied variables when complex template filter is received', () => {
    expect(applyVariablesInFilters(multiVariablesTemplateFilter, mockedTableMetadata, scopedVars)).toMatchSnapshot();
  });
  it('should return empty filter, when the variable in clause has "All" selected, and clause has no parent', () => {
    expect(applyVariablesInFilters(allValFilter, mockedTableMetadata, scopedVars)).toEqual({});
  });
  it('should return filter with single clause, when the filter has "and" clause, and variable in child clause has "All" selected', () => {
    expect(applyVariablesInFilters(allInAndFilter, mockedTableMetadata, scopedVars)).toMatchSnapshot();
  });

  it('should return filter with "and" that has multiple clauses, when the "and" filter with more than 2 clauses received, and one of them has "All" selected', () => {
    expect(applyVariablesInFilters(allInMultipleClauseAndFilter, mockedTableMetadata, scopedVars)).toMatchSnapshot();
  });
  it('should return empty filter, when the "or" filter with multiple clause received, and one of its children has "All" selected', () => {
    expect(applyVariablesInFilters(allInOrFilter, mockedTableMetadata, scopedVars)).toEqual({});
  });
  it('should return filter with one clause, when complex filter with "All" is received', () => {
    expect(applyVariablesInFilters(complexAllFilter, mockedTableMetadata, scopedVars)).toMatchSnapshot();
  });
  it('should return filter with applied custom "All" value', () => {
    expect(applyVariablesInFilters(customAllFilter, mockedTableMetadata, scopedVars)).toMatchSnapshot();
  });
  it('should throw error when variables with "All" value is used in invalid format', () => {
    expect(() => applyVariablesInFilters(invalidTemplateAllFilter, mockedTableMetadata, scopedVars)).toThrowError(
      'The provided template "${VLUNAME}test${VLUNAME}" does not match the expected format for variables with value "All"'
    );
  });
  it('should throw error when variable is not found in grafana', () => {
    expect(() => applyVariablesInFilters(varNotFoundInDashboardFilter, mockedTableMetadata, scopedVars)).toThrowError(
      'There is no variable information for "$VarNotFoundInDashboard"'
    );
  });
  it('should correctly handle variables applying', () => {
    expect(applyVariablesInFilters(mgtstatFilter, mockedTableMetadata, scopedVars)).toMatchSnapshot();
    expect(applyVariablesInFilters(swpstatFilter, mockedTableMetadata, scopedVars)).toMatchSnapshot();
    expect(applyVariablesInFilters(frmsusepctFilter, mockedTableMetadata, scopedVars)).toMatchSnapshot();
  });
});

describe('handleClause', () => {
  it('should return unchanged clause, when template contains no variables', () => {
    expect(
      handleClause(
        noVarInTemplateFilter.nonAggregated.clause,
        mockedTableMetadata.columns.DB2ID as ColumnMetadata,
        scopedVars
      )
    ).toEqual({ clause: noVarInTemplateFilter.nonAggregated.clause });
  });

  it('should return clause with interpolated variables and formatted value, when template contains variables', () => {
    expect(
      handleClause(singleVarFilter.clause, mockedTableMetadata.columns.DB2ID as ColumnMetadata, scopedVars)
    ).toMatchSnapshot();
    expect(
      handleClause(mgtstatClause, mockedTableMetadata.columns.MGTSTAT as ColumnMetadata, scopedVars)
    ).toMatchSnapshot();
    expect(
      handleClause(swpstatClause, mockedTableMetadata.columns.SWPSTAT as ColumnMetadata, scopedVars)
    ).toMatchSnapshot();
    expect(
      handleClause(frmsusepctClause, mockedTableMetadata.columns.FRMSUSEPCT as ColumnMetadata, scopedVars)
    ).toMatchSnapshot();
  });

  it('should throw, when template contains variable that is not found in dashboard', () => {
    expect(() => {
      handleClause(
        varNotFoundInDashboardFilter.nonAggregated.clause,
        mockedTableMetadata.columns.JOBNAME as ColumnMetadata,
        scopedVars
      );
    }).toThrow();
  });
});

describe('handleClauseWithMultipleValues', () => {
  it('should return "or" filter, when passed clause and string values', () => {
    expect(handleClauseWithMultipleValues(multiValueClause, ['USSDSYS', 'IMSRGNS', 'SYSSTC'])).toMatchSnapshot();
  });
});
