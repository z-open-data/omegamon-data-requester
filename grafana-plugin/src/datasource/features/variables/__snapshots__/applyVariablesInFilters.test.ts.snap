// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`applyVariablesInFilters should correctly handle variables applying 1`] = `
{
  "aggregated": undefined,
  "nonAggregated": {
    "clause": {
      "columnId": "MGTSTAT",
      "operator": "=",
      "userDefinedValue": "Unknown",
    },
  },
}
`;

exports[`applyVariablesInFilters should correctly handle variables applying 2`] = `
{
  "aggregated": undefined,
  "nonAggregated": {
    "clause": {
      "columnId": "SWPSTAT",
      "operator": "=",
      "userDefinedValue": "OutLSW",
    },
  },
}
`;

exports[`applyVariablesInFilters should correctly handle variables applying 3`] = `
{
  "aggregated": undefined,
  "nonAggregated": {
    "clause": {
      "columnId": "FRMSUSEPCT",
      "operator": "=",
      "userDefinedValue": "43.4",
    },
  },
}
`;

exports[`applyVariablesInFilters should return filter with "and" that has multiple clauses, when the "and" filter with more than 2 clauses received, and one of them has "All" selected 1`] = `
{
  "aggregated": undefined,
  "nonAggregated": {
    "and": [
      {
        "clause": {
          "columnId": "DB2ID",
          "operator": "=",
          "userDefinedValue": "OBC4",
        },
      },
      {
        "clause": {
          "columnId": "SVCCLASS",
          "operator": "=",
          "userDefinedValue": "USSDSYS",
        },
      },
    ],
  },
}
`;

exports[`applyVariablesInFilters should return filter with applied custom "All" value 1`] = `
{
  "aggregated": undefined,
  "nonAggregated": {
    "or": [
      {
        "clause": {
          "columnId": "DB2SUBSYSTEM",
          "operator": "=",
          "userDefinedValue": "OCA5",
        },
      },
      {
        "clause": {
          "columnId": "DB2ID",
          "operator": "=",
          "userDefinedValue": "OBC4",
        },
      },
      {
        "clause": {
          "columnId": "SVCCLASS",
          "operator": "=",
          "userDefinedValue": "USSDSYS",
        },
      },
    ],
  },
}
`;

exports[`applyVariablesInFilters should return filter with correctly applied variables when complex template filter is received 1`] = `
{
  "aggregated": undefined,
  "nonAggregated": {
    "or": [
      {
        "clause": {
          "columnId": "JOBNAME",
          "operator": "=",
          "userDefinedValue": "fooRMFGATbarUSSDSYSbazXCFAS",
        },
      },
      {
        "clause": {
          "columnId": "JOBNAME",
          "operator": "=",
          "userDefinedValue": "fooRMFGATbarUSSDSYSbazWLM",
        },
      },
      {
        "clause": {
          "columnId": "JOBNAME",
          "operator": "=",
          "userDefinedValue": "fooRMFGATbarUSSDSYSbazRMFGAT",
        },
      },
      {
        "clause": {
          "columnId": "JOBNAME",
          "operator": "=",
          "userDefinedValue": "fooRMFGATbarIMSRGNSbazXCFAS",
        },
      },
      {
        "clause": {
          "columnId": "JOBNAME",
          "operator": "=",
          "userDefinedValue": "fooRMFGATbarIMSRGNSbazWLM",
        },
      },
      {
        "clause": {
          "columnId": "JOBNAME",
          "operator": "=",
          "userDefinedValue": "fooRMFGATbarIMSRGNSbazRMFGAT",
        },
      },
      {
        "clause": {
          "columnId": "JOBNAME",
          "operator": "=",
          "userDefinedValue": "fooRMFGATbarSYSSTCbazXCFAS",
        },
      },
      {
        "clause": {
          "columnId": "JOBNAME",
          "operator": "=",
          "userDefinedValue": "fooRMFGATbarSYSSTCbazWLM",
        },
      },
      {
        "clause": {
          "columnId": "JOBNAME",
          "operator": "=",
          "userDefinedValue": "fooRMFGATbarSYSSTCbazRMFGAT",
        },
      },
    ],
  },
}
`;

exports[`applyVariablesInFilters should return filter with one clause, when complex filter with "All" is received 1`] = `
{
  "aggregated": undefined,
  "nonAggregated": {
    "clause": {
      "columnId": "JOBNAME",
      "operator": "=",
      "userDefinedValue": "RMFGAT",
    },
  },
}
`;

exports[`applyVariablesInFilters should return filter with single clause, when the filter has "and" clause, and variable in child clause has "All" selected 1`] = `
{
  "aggregated": undefined,
  "nonAggregated": {
    "clause": {
      "columnId": "DB2ID",
      "operator": "=",
      "userDefinedValue": "OBC4",
    },
  },
}
`;

exports[`applyVariablesInFilters should return processed filter with nonAggregated field 1`] = `
{
  "aggregated": undefined,
  "nonAggregated": {
    "or": [
      {
        "clause": {
          "columnId": "JOBNAME",
          "operator": "=",
          "userDefinedValue": "RMFGAT",
        },
      },
      {
        "and": [
          {
            "or": [
              {
                "clause": {
                  "columnId": "SVCCLASS",
                  "operator": "=",
                  "userDefinedValue": "USSDSYS",
                },
              },
              {
                "clause": {
                  "columnId": "SVCCLASS",
                  "operator": "=",
                  "userDefinedValue": "IMSRGNS",
                },
              },
              {
                "clause": {
                  "columnId": "SVCCLASS",
                  "operator": "=",
                  "userDefinedValue": "SYSSTC",
                },
              },
            ],
          },
          {
            "or": [
              {
                "clause": {
                  "columnId": "STEPNAME",
                  "operator": "=",
                  "userDefinedValue": "XCFAS",
                },
              },
              {
                "clause": {
                  "columnId": "STEPNAME",
                  "operator": "=",
                  "userDefinedValue": "WLM",
                },
              },
              {
                "clause": {
                  "columnId": "STEPNAME",
                  "operator": "=",
                  "userDefinedValue": "RMFGAT",
                },
              },
            ],
          },
        ],
      },
    ],
  },
}
`;

exports[`applyVariablesInFilters should return unchanged filter when there is no variable defined in "userDefinedValue" 1`] = `
{
  "aggregated": undefined,
  "nonAggregated": {
    "clause": {
      "columnId": "DB2ID",
      "operator": "=",
      "userDefinedValue": "OBC4",
    },
  },
}
`;

exports[`handleClause should return clause with interpolated variables and formatted value, when template contains variables 1`] = `
{
  "clause": {
    "columnId": "DB2ID",
    "operator": "=",
    "userDefinedValue": "OBC4",
  },
}
`;

exports[`handleClause should return clause with interpolated variables and formatted value, when template contains variables 2`] = `
{
  "clause": {
    "columnId": "MGTSTAT",
    "operator": "=",
    "userDefinedValue": "Unknown",
  },
}
`;

exports[`handleClause should return clause with interpolated variables and formatted value, when template contains variables 3`] = `
{
  "clause": {
    "columnId": "SWPSTAT",
    "operator": "=",
    "userDefinedValue": "OutLSW",
  },
}
`;

exports[`handleClause should return clause with interpolated variables and formatted value, when template contains variables 4`] = `
{
  "clause": {
    "columnId": "FRMSUSEPCT",
    "operator": "=",
    "userDefinedValue": "43.4",
  },
}
`;

exports[`handleClauseWithMultipleValues should return "or" filter, when passed clause and string values 1`] = `
{
  "or": [
    {
      "clause": {
        "columnId": "SVCCLASS",
        "operator": "=",
        "userDefinedValue": "USSDSYS",
      },
    },
    {
      "clause": {
        "columnId": "SVCCLASS",
        "operator": "=",
        "userDefinedValue": "IMSRGNS",
      },
    },
    {
      "clause": {
        "columnId": "SVCCLASS",
        "operator": "=",
        "userDefinedValue": "SYSSTC",
      },
    },
  ],
}
`;
